### PR TITLE
[2] Implemented Record Transformations

### PR DESCRIPTION
**Pull request by aivantg**
## Implemented Record Transformations 

Implemented Record Transformations so that Airtable column names are transformed to and from conventional JS attribute camel case format when they are retrieved and sent. 

## Motivation

There's a lot of places where we have really useless conversions from capitalized variable names to lowercase variable names. I think it would look a lot more natural and be easier to develop if the column names were what you expected them to be. 

By transforming the data to and from a more conventional format at the lowest level of abstractions, we can work with records in the most natural javascript way without having to deal with airtable column names.

[More Info on Notion](https://www.notion.so/calblueprint/PP-Power-Airtable-Client-Side-6a1b6734af294ef88609a6d6d256ca3d#f543574a2e5f4218ac9fb0ab8e5adda2)

### Next Steps

Clean up the code to remove redundant state setting! 

### How to Review

Most important code changes are in `airtable.js`. Everything is just refactoring to use new column names

### Related PRs

Branches off of #59 

### Tests Performed, Edge Cases

We should test each corner of the app thoroughly for bugs with this PR as so many core pieces have changed

[//]: # (If you made changes/additions to `airtable.js` or `request.js`,) 
[//]: # (make sure you update/add [at least 1] corresponding unit tests in)
[//]: # (`airtable.spec.js` or `request.spec.js`)

CC: @aivantg @dfangshuo

### CODE REVIEW

File: src/screens/general/UserProfilePage.js [hunk 1]
```diff
@@ -49,22 +49,24 @@ class UserProfilePage extends React.Component {
...
-      'RECORD ID (for dev)': id,
-      Email: email,
-      'Phone Number': phoneNumber,
-      Name: name,
-      'User Login': userLoginID,
-      City: city,
-      Street: street,
-      State: state,
-      Zipcode: zipCode
+      recordIdforDev: id,
```
Comment (dfangshuo) [inline src/screens/general/UserProfilePage.js]: hmm should camelCase convention dictate that for also be capitalized? this is a perhaps a change that needs to be made in the airtable schema generator?

File: src/screens/onboarding/BasicInfo.js [hunk 2]
```diff
@@ -34,13 +34,13 @@ class BasicInfo extends React.Component {
       if (userId) {
         const fullName = `${fname} ${lname}`;
```
Comment (dfangshuo) [inline src/screens/onboarding/BasicInfo.js]: maybe assign this variable name and then just pass in `name` rather than `name: fullName` to conform with how this is done above?

File: src/components/NavBar.js [hunk 3]
```diff
@@ -17,7 +16,7 @@ class NavBar extends React.PureComponent {
     const { person, credentials, isNavBarVisible } = this.props;
     // TODO: Holy crap so apparently the "ID" column on airtable is their NAME not the record ID.
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/components/NavBar.js]: You might be right. My opinion is/was that the Primary Key, left-most column be something like their name or something usable, because as far as I can see, it's main purpose is to give a sort of 'toString' method for the record when referencing it in other records. _Then_ we should also have an "ID" column that is the record's actual ID. The Primary key column should NOT be named "ID"

I don't think there's any harm to performance by letting our primary key be the name because I'm sure airtable optimizes for record ID lookups anyway. More info here, confirming that the primary key column is primarily for visual usability: https://blog.airtable.com/the-primary-column/

File: src/components/progressbar.js [hunk 4]
```diff
@@ -10,6 +10,7 @@ function colorStep(activeStep, step) {
   return '';
 }
 
+// TODO: convert to a react component
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/components/progressbar.js]: It's like, 99% of the way to a proper react component, but a proper functional react component takes in a props object as a parameter while this takes in a single "step" parameter. Seems arbitrary, but when it's time to _use_ this component, we now have to do something like: 
```
<div>
  {ProgressBar(3)}
</div>
```
instead of the more conventional 
```
<div>
  <ProgressBar step=3 />
</div>
```

i think we should prefer the second one.

File: src/lib/adminUtils.js [hunk 5]
```diff
@@ -18,10 +17,9 @@ const getAdminTable = async owner => {
 // Can't the admin dashboard just deal in IDs instead of the actual objects?
 const getOwnersFromProjectGroup = async projectGroupId => {
   const projectGroup = await getProjectGroupById(projectGroupId);
-  const { Owner } = projectGroup;
 
   const ownersObjects = await Promise.all(
-    Owner.map(ownerId => getOwnerById(ownerId))
+    projectGroup.owner.map(ownerId => getOwnerById(ownerId))
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/adminUtils.js]: I think eslint only cares about desctructuring props and state? 

Did not get yelled at and didn't change any rules as far as I know of (and honestly I think it's overkill to constantly destructure everything, my feelings are mixed). That being said if we wanna stay consistent I can come back and change this to use destrucutring

File: src/lib/airtable/airtable.js [hunk 6]
```diff
@@ -10,24 +10,57 @@
...
 Airtable.configure({
   endpointUrl: ENDPOINT_URL,
-  apiKey: key
+  apiKey
 });
 
 const base = Airtable.base(BASE_ID);
 
+// Transformation Utilities
+const cleanTableName = name => name.replace(/\(|\)|\s/g, '');
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/airtable/airtable.js]: Yes! I will add comments

File: src/lib/airtable/airtable.js [hunk 7]
```diff
@@ -10,24 +10,57 @@
...
+
+const transformRecordFromAirtableFormat = (record, tableName) => {
+  const table = cleanTableName(tableName);
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/airtable/airtable.js]: it's honestly just because of the structure of `schema.js` that it uses "clean table names" as the key instead of the actual table name. In hindsight i think I might change this to be more consistent

File: src/lib/airtable/airtable.js [hunk 8]
```diff
@@ -10,24 +10,57 @@
...
-  apiKey: key
+  apiKey
 });
 
 const base = Airtable.base(BASE_ID);
 
+// Transformation Utilities
+const cleanTableName = name => name.replace(/\(|\)|\s/g, '');
+
+const transformRecordFromAirtableFormat = (record, tableName) => {
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/airtable/airtable.js]: yesss, great idea. I couldn't think of short names so i went super verbose instead 😂  I like those!

File: src/lib/airtable/airtable.js [hunk 9]
```diff
@@ -10,24 +10,57 @@
...
+  const table = cleanTableName(tableName);
+  // Invert columns so you can look up by Airtable Column Name
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/airtable/airtable.js]: Sure, i'll see if I can write up a more useful comment. But in short, the `Columns` object in `schema.js` has the js formatted name as a key and the airtable format as a value. Inverting the object switches that so we can look up the js formatted name using the airtable formatted name.

File: src/lib/airtable/airtable.js [hunk 10]
```diff
@@ -10,24 +10,57 @@
...
+  const columns = {};
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/airtable/airtable.js]: 👍

File: src/lib/airtable/airtable.js [hunk 11]
```diff
@@ -10,24 +10,57 @@
...
+  Object.keys(record).forEach(origName => {
+    const jsFormattedName = columns[origName];
+    newRecord[jsFormattedName] = record[origName];
+  });
+
+  return newRecord;
+};
+
+const transformRecordToAirtableFormat = (record, tableName) => {
+  const table = cleanTableName(tableName);
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/airtable/airtable.js]: You're right, we don't really need to clean it, it's just here because when I inserted keys into the column table, I cleaned it so it would be easy to do something like `Columns.UserLogin.id` instead of `Columns['User Login'].id` but now that we should never be looking up column names in our code anyway, it's probably not a big deal and i could get rid of cleaning names

File: src/lib/airtable/schema.json [hunk 12]
```diff
@@ -151,8 +151,7 @@
       "ID",
       "Location",
       "Time"
-    ],
-    "lookupFields": ["Project Group"]
+    ]
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/airtable/schema.json]: it's basically going from an old version of how my schema generator worked to a new version. `schema.json` is now a fully generated file that just exists as reference for us. `schemaMeta.json` is the input file to the generator that specifies lookup fields and all

File: src/lib/airtable/schemaMeta.json [hunk 13]
```diff
@@ -0,0 +1,8 @@
+{
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/airtable/schemaMeta.json]: details are in the repo's readme I think!

[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/airtable/schemaMeta.json]: I can make them clearer though if it's still confusing

File: src/lib/onboarding/onboardingUtils.js [hunk 14]
```diff
@@ -9,10 +9,11 @@ import {
 const DEFAULT_NUM_RETRIES = 3;
 
 // Returns created person's ID
+// TODO: maybe have this function accept an object to...simplify
```
[the remark this answers is not in the document]
Comment (aivantg) [reply src/lib/onboarding/onboardingUtils.js]: Yeah, I agree in general and was going to suggest this for all of our `create` functions in record.js, and my suggestion was that we go towards like, named parameters too. But I realized that Javascript doesn't _have_ named parameters, objects as parameters is their solution for that lol. 

We can talk about this more in person because it's a super nitpick thing that i'm still very unsure about

File: src/screens/general/UserProfilePage.js (hunk 1 above)
Comment (aivantg) [reply src/screens/general/UserProfilePage.js]: hmm...good point. But I let it slide initially because I want this column name to just be changed to "ID" instead of "RECORD ID (for dev)" But yeah looks like i should take the parenthesis out before i split by spaces. will note for next version of the schema generator

File: src/screens/onboarding/BasicInfo.js (hunk 2 above)
Comment (aivantg) [reply src/screens/onboarding/BasicInfo.js]: sure sure

File: src/components/NavBar.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (dfangshuo) [reply src/components/NavBar.js]: makes sense! flagging this as an issue in our notion

[the remark this answers is not in the document]
Comment (dfangshuo) [reply src/components/NavBar.js]: ^ never mind saw that you already did it

File: src/lib/adminUtils.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (dfangshuo) [reply src/lib/adminUtils.js]: if eslint is chill with it i dont have any issues with it lmao
